src/NHibernate/Linq/Visitors/TransparentIdentifierRemovingExpressionVisitor.cs

```diff
@@ -0,0 +1,117 @@
+﻿// Copyright (c) rubicon IT GmbH, www.rubicon.eu
+//
+// See the NOTICE file distributed with this work for additional information
+// regarding copyright ownership.  rubicon licenses this file to you under 
+// the Apache License, Version 2.0 (the "License"); you may not use this 
+// file except in compliance with the License.  You may obtain a copy of the 
+// License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software 
+// distributed under the License is distributed on an "AS IS" BASIS, WITHOUT 
+// WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  See the 
+// License for the specific language governing permissions and limitations
+// under the License.
+// 
+
+using System.Collections.Generic;
+using System.Linq;
+using System.Linq.Expressions;
+using System.Reflection;
+using Remotion.Linq;
+using Remotion.Linq.Clauses.Expressions;
+using Remotion.Linq.Parsing;
+using MemberBinding = Remotion.Linq.Parsing.ExpressionVisitors.MemberBindings.MemberBinding;
+
+namespace NHibernate.Linq.Visitors
+{
+	// Copied from Relinq and added a fallback for comparing two member info by DeclaringType and Name
+	/// <summary>
+	/// Replaces expression patterns of the form <c>new T { x = 1, y = 2 }.x</c> (<see cref="MemberInitExpression"/>) or 
+	/// <c>new T ( x = 1, y = 2 ).x</c> (<see cref="NewExpression"/>) to <c>1</c> (or <c>2</c> if <c>y</c> is accessed instead of <c>x</c>).
+	/// Expressions are also replaced within subqueries; the <see cref="QueryModel"/> is changed by the replacement operations, it is not copied. 
+	/// </summary>
+	internal sealed class TransparentIdentifierRemovingExpressionVisitor : RelinqExpressionVisitor
+	{
+		public static Expression ReplaceTransparentIdentifiers(Expression expression)
+		{
+			Expression expressionBeforeRemove;
+			Expression expressionAfterRemove = expression;
+
+			// Run again and again until no replacements have been made.
+			do
+			{
+				expressionBeforeRemove = expressionAfterRemove;
+				expressionAfterRemove = new TransparentIdentifierRemovingExpressionVisitor().Visit(expressionAfterRemove);
+			} while (expressionAfterRemove != expressionBeforeRemove);
+
+			return expressionAfterRemove;
+		}
+
+		private TransparentIdentifierRemovingExpressionVisitor()
+		{
+		}
+
+		protected override Expression VisitMember(MemberExpression memberExpression)
+		{
+			var memberBindings = GetMemberBindingsCreatedByExpression(memberExpression.Expression);
+			if (memberBindings == null)
+				return base.VisitMember(memberExpression);
+
+			var matchingAssignment = memberBindings
+			                         .Where(binding => binding.MatchesReadAccess(memberExpression.Member))
+			                         .LastOrDefault();
+
+			// Added logic: In some cases (e.g OData), the member can be from a different derived class, in such case
+			// we need to check the member DeclaringType instead of ReflectedType
```

Comment:
It's all good for now, but probably one of 2 things should happen:

1. This fixed in OData provider so it generates the correct expression.
2. This fixed in ReLinq so we do not copy code from there and just use fixed version.

Reply:
> but probably one of 2 things should happen:

I do agree and [created an issue](https://github.com/OData/WebApi/issues/2108) for ODAta, which I thnik is where the issue should be fixed.